signac/core/collection_api.py

```diff
@@ -0,0 +1,491 @@
+        self.load()
+        return reversed(self._data)
+
+    def __iadd__(self, iterable):
+        self.load()
+        with self._safe_sync():
+            self._data += [self.from_base(data=value, parent=self) for value in iterable]
+            self.sync()
+
+    def __call__(self):
+        self.load()
+        return self.to_base()
+
+    def __eq__(self, other):
+        if isinstance(other, type(self)):
+            return self() == other()
+        else:
+            return self() == other
+
+    def __repr__(self):
+        return repr(self())
+
+    def __str__(self):
+        return str(self())
+
+    def insert(self, index, item):
+        self.load()
+        with self._safe_sync():
+            with self._suspend_sync():
+                self._data.insert(index, self.from_base(data=item, parent=self))
+            self.sync()
+
+    def append(self, item):
+        self.load()
+        with self._safe_sync():
+            with self._suspend_sync():
+                self._data.append(self.from_base(data=item, parent=self))
+            self.sync()
+
+    def extend(self, iterable):
+        self.load()
+        with self._safe_sync():
+            with self._suspend_sync():
+                self._data.extend([self.from_base(data=value, parent=self) for value in iterable])
+            self.sync()
+
+    def remove(self, item):
+        self.load()
+        with self._safe_sync():
+            with self._suspend_sync():
+                self._data.remove(self.from_base(data=item, parent=self))
+            self.sync()
+
+    def clear(self):
+        with self._safe_sync():
+            self._data = []
+            self.sync()
+
+
+class JSONCollection(SyncedCollection):
+
+    def __init__(self, **kwargs):
+        filename = kwargs.pop('filename', None)
+        self._filename = None if filename is None else os.path.realpath(filename)
+        self._write_concern = kwargs.pop('write_concern', True)
+        super().__init__(**kwargs)
+
+    def _load(self):
+        try:
+            with open(self._filename, 'rb') as file:
+                blob = file.read()
+                return json.loads(blob.decode())
+        except IOError as error:
+            if error.errno == errno.ENOENT:
+                return None
+
+    def _sync(self):
+        data = self.to_base()
+        # Serialize data:
+        blob = json.dumps(data).encode()
+
+        if self._write_concern:
+            dirname, filename = os.path.split(self._filename)
+            fn_tmp = os.path.join(dirname, '._{uid}_{fn}'.format(
+                uid=uuid.uuid4(), fn=filename))
+            with open(fn_tmp, 'wb') as tmpfile:
+                tmpfile.write(blob)
+            os.replace(fn_tmp, self._filename)
+        else:
+            with open(self._filename, 'wb') as file:
+                file.write(blob)
+
+
+class JSONDict(JSONCollection, SyncedAttrDict):
+    def __init__(self, filename=None, data=None, parent=None, write_concern=False):
```

Comment:
I would prefer the argument order used above:
```suggestion
    def __init__(self, data=None, filename=None, parent=None, write_concern=False):
```